### PR TITLE
Improve MazeGen

### PR DESCRIPTION
**Pull request by jensj12**
- May now generate mazes that require you to pickup a key
 - Only places blocking doors after a key has been placed
 - Adds difficulty setting
 - Places the exit as far away as possible from the player start position, with a bias to place it behind blocking doors
 - Use helper functions to make main code more readable

### CODE REVIEW

File: MeesGame/Gameplay/Generator/MazeGenerator.cs [hunk 1]
```diff
@@ -12,16 +12,57 @@ class MazeGenerator
...
+            double loopChance = 0, splitChance = 0, doorChance = 0;
+            int numKeys = 0;
+            difficulty = MathHelper.Clamp(difficulty, 1, 5);
+            switch (difficulty)
+            {
+                case 1:
+                    loopChance = 0.05;
+                    splitChance = 0.03;
+                    doorChance = 0;
+                    numKeys = 3;
```
Comment (jorisBarkema) [inline MeesGame/Gameplay/Generator/MazeGenerator.cs:33]: Why place keys if there aren't going to be any doors?

File: MeesGame/Gameplay/Generator/MazeGenerator.cs [hunk 2]
```diff
@@ -31,15 +72,29 @@ public static TileField GenerateMaze(int numRows = DEFAULT_NUM_ROWS, int numCols
...
             Point start = new Point(random.Next(tiles.Columns / 2 - 1) * 2 + 1, random.Next(tiles.Rows / 2 - 1) * 2 + 1);
+            bestExit = start;
+            int totalNodesToDo = (tiles.Rows - 1) * (tiles.Columns - 1) / 4;
+            int[] keyNodes = new int[numKeys];
+            for (int i = 0; i < numKeys; i++)
+            {
+                do
+                {
+                    keyNodes[i] = (random.Next(totalNodesToDo) + random.Next(totalNodesToDo)) / 2; //try to get it halfway
+                } while (isDuplicate(keyNodes, i));
```
Comment (jorisBarkema) [inline MeesGame/Gameplay/Generator/MazeGenerator.cs:89]: I looked at the isDuplicate method, it seems to me you should be able to do this using somethign like this

`keyNodes.GetRange(0, i).Contains(keyNodes[i])`

**Review by jorisBarkema — COMMENTED** [review]
Looks good.

File: MeesGame/Gameplay/Generator/MazeGenerator.cs (hunk 2 above)
Comment (jensj12) [reply MeesGame/Gameplay/Generator/MazeGenerator.cs:89]: An array doesn't have a GetRange or Contains functions. If you can find something else that works, please let me know.

File: MeesGame/Gameplay/Generator/MazeGenerator.cs (hunk 1 above)
Comment (jensj12) [reply MeesGame/Gameplay/Generator/MazeGenerator.cs:33]: Loops are doors, but I've added another difficulty parameter for that.

**Comment by jorisBarkema** [comment]
I like the idea of the loopDoorChance, but wouldn't it be better to use it for other obstacles, because we don't have any of those yet? I don't see a real reason why they have to be doors in this case.

**Comment by jensj12** [comment]
There are no other obstacles yet. Other obstacles may even need to have a seperate chance variable.

**Comment by jorisBarkema** [comment]
In that case maybe it would be better to keep the code for later, because I do like using loops to place obstacles, but just place walls for now. I think it would be confusing/misleading to place doors you can't open.

**Comment by jensj12** [comment]
You can open them... There are keys on the level for that reason (you should know because you pointed it out in the code).

**Comment by jorisBarkema** [comment]
Yes but when a key is placed placeDoors = true so I thought this was to make implementing matching specific doors corresponding with specific keys in the future easier. If every key fits every door, it's great.

**Comment by jensj12** [comment]
LoopDoors don't block paths, they only add more after a key has been collected. Therefore I don't check for placed keys there. PlaceDoors (or keysPlaced) will become an array once colored keys and doors are added to the generator. That will be for another PR. (if you want to do so, just claim the task)

**Comment by jorisBarkema** [comment]
Yeah I know how it works, my only point was once we add colored keys they'll match with the doors from placeDoors, not the ones in loops. So it's good, placing more doors in loops makes the level more difficult, but I think once there are more obstacles we can place we should consider placing them in the way you're placing the loopdoors now.

**Review by jorisBarkema — APPROVED** [review]

**Pull request merged**